doc/testing.md

```diff
@@ -4,8 +4,94 @@
 
 If you are unit testing your application you should stub all external services.
 
+### Interpreter
+
+If you are using tapir do describe your API and run using some server interpreter then you might want to test your
+endpoints definitions (like validations, decoding) as well as your custom server options (like interceptors, failure handlers) while 
+stubbing the actual response of your server logic.
+
+Add the dependency:
+
+```scala
+"com.softwaremill.sttp.tapir" %% "tapir-sttp-stub-server" % "@VERSION@"
+```
+
+Let's assume you are using Akka interpreter, then given the following endpoint:
+
+```scala mdoc:invisible
+import org.scalatest.flatspec.AsyncFlatSpec
+import org.scalatest.matchers.should.Matchers
+import sttp.client3._
+import sttp.model.StatusCode
+import sttp.monad.FutureMonad
+import sttp.tapir._
+import sttp.tapir.server.ServerEndpoint.Full
+import sttp.tapir.server.akkahttp.AkkaHttpServerOptions
+import sttp.tapir.server.interceptor.exception.ExceptionContext
+import sttp.tapir.server.interceptor.{CustomInterceptors, ValuedEndpointOutput}
+import sttp.tapir.server.stub.TapirStubInterpreter
+import scala.concurrent.Future
+```
+
+```scala mdoc:silent
+val greetUser: Full[String, String, Unit, String, String, Any, Future] = endpoint.get
+  .in("api" / "users" / "greet")
+  .securityIn(auth.bearer[String]())
+  .out(stringBody)
+  .errorOut(stringBody)
+  .serverSecurityLogic(token =>
+    Future.successful {
+      (if (token == "secret-password") Right("user123") else Left("unauthorized"))
+    }
+  )
+  .serverLogic(user => _ => Future.successful(Right(s"hello $user")))
+```
+
+And the following `AkkaHttpServerOptions`:
+
+```scala mdoc:silent
+val options: CustomInterceptors[Future, AkkaHttpServerOptions] = AkkaHttpServerOptions.customInterceptors
+  .exceptionHandler((ctx: ExceptionContext) =>
+    Some(ValuedEndpointOutput(stringBody.and(statusCode), (s"failed due to ${ctx.e.getMessage}", StatusCode.InternalServerError)))
+  )
+```
+
+You can write an interpreter stub for your endpoint. 
+The `backend()` method returns `SttpBackend` which runs all requests against interpreter and can be used in tests:
+
+```scala mdoc:silent
+class MySpec extends AsyncFlatSpec with Matchers {
+  
+  val stub = TapirStubInterpreter[Future, Any, AkkaHttpServerOptions](options.serverLog(None), new FutureMonad())
+    .forServerEndpoint(greetUser)
```

Comment:
we're usually using `from...` so this would be `fromServerEndpoint`

Reply:
Ah I see that you can call `.for(...).return(...)` multiple times. Then `from` doesn't make sense. But maybe we could stick with `.whenEndpoint(...).thenSuccess(...)`? I'm not a fan of `return`, `respond` would be better as it matches http vocabulary, but maybe we can just remove it ;)

Reply:
Alternatively, we could do `.for(...).respondWith(...)`, but that's longer

Reply:
I'll change it to `for..` and `respond` then 👍